portfolio/src/main/java/com/google/sps/servlets/BookDataServlet.java

```diff
@@ -40,9 +55,29 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String book = request.getParameter("book");
-    int currentVotes = bookVotes.containsKey(book) ? bookVotes.get(book) : 0;
-    bookVotes.put(book, currentVotes + 1);
+    long currentVotes = 1;
+
+    Query query = new Query("BookVotes");
+    PreparedQuery results = datastore.prepare(query);
+
+    // Search if the book is already in the list
```

Comment:
You can use a filter directly in the Query (and then DatastoreService will filter for you):
  Query query = new Query("BookVotes").setFilter(PropertyFilter.eq("name", book));

[Hint: when using an AppEngine class, consider reading its tutorial - in the case of Query, it is http://cloud/datastore/docs/concepts/queries#datastore-datastore-basic-query-java ]